File: 백준/Platinum/11003. 최솟값 찾기/최솟값 찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

class Main {
    static Node[] seq;
    static int N, L;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        L = Integer.parseInt(st.nextToken());
        seq = new Node[N];

        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < N; i++) {
            seq[i] = new Node(i, Integer.parseInt(st.nextToken()));
        }

        StringBuilder sb = new StringBuilder();
        Deque<Node> deque = new ArrayDeque<>(L);

        for (int e = 0; e < N; e++) {
            Node next = seq[e];

            while (!deque.isEmpty() && deque.getLast().value > next.value) {
                deque.removeLast();
            }

            deque.addLast(next);

            if (deque.getFirst().index < e - L + 1) {
                deque.removeFirst();
            }

            sb.append(deque.getFirst().value).append(' ');
        }

        System.out.println(sb);
        br.close();
    }
}

class Node {
    int index;
    int value;

    Node(int index, int value) {
        this.index = index;
        this.value = value;
    }
}

File: 백준/Platinum/11003. 최솟값 찾기/최솟값 찾기.kt
import java.util.*

var cap = 1
lateinit var tree: IntArray

fun main() {
    val br = System.`in`.bufferedReader()
    var st = StringTokenizer(br.readLine())

    val n = st.nextToken().toInt()
    val l = st.nextToken().toInt()

    while (cap < n) cap *= 2

    st = StringTokenizer(br.readLine())
    tree = IntArray(cap shl 1)

    for (i in 0 until n) {
        tree[cap + i] = st.nextToken().toInt()
    }

    for (node in cap - 1 downTo 1) {
        val child = node shl 1

        tree[node] = minOf(tree[child], tree[child + 1])
    }

    val sb = StringBuilder()

    for (e in 0 until n) {
        val s = maxOf(0, e - l + 1)

        sb.append(query(s, e)).append(' ')
    }

    println(sb)
}

fun query(qs: Int, qe: Int): Int {
    var ret = Int.MAX_VALUE
    var s = cap + qs
    var e = cap + qe

    while (s < e) {
        if (s and 1 == 1) {
            ret = minOf(ret, tree[s++])
        }

        if (e and 1 == 0) {
            ret = minOf(ret, tree[e--])
        }

        ret = minOf(ret, tree[s], tree[e])
        s = s ushr 1
        e = e ushr 1
    }

    if (s == e) {
        ret = minOf(ret, tree[s])
    }

    return ret
}